JavaScript/Next-js.md

```diff
@@ -0,0 +1,221 @@
+import Head from 'next/head';
+
+import withAuth from '../higherOrderComponents/withAuth';
+
+const Layout = (props) => {
+  return (
+    <Fragment>
+      <Head>
+        <link
+          href="https://fonts.googleapis.com/css?family=Roboto"
+          rel="stylesheet"
+          key="google-font-cabin"
+        />
+      </Head>
+
+      <div id="portal"></div>
+
+      <main>
+        <div className="container">{props.children}</div>
+      </main>
+
+      <style global jsx>{`
+        body {
+          font-family: 'Roboto', sans-serif;
+        }
+      `}</style>
+    </Fragment>
+  );
+};
+
+export default withAuth(Layout);
+```
+
+Here is an example of how we can use `Layout`:
+
+```jsx
+// /pages/index.tsx
+import React from 'react';
+// ...
+import Layout from '../components/layouts/Layout';
+
+const Index = () => {
+  // ...
+  return <Layout>Home Page content</Layout>;
+};
+export default Index;
+```
+
+### Higher order components
+
+From react docs HOC-s are **functions which take components and returns a new component**. They are used when we need to re-use some component logic.
+
+### WithAuth
+
+`withAuth` is simple HOC which checks if the user is authenticated, if not user is redirected to the login page.
+
+```jsx
+// /components/higherOrderComponents/withAuth.tsx
+import * as React from 'react';
+import { useEffectOnce } from 'react-use';
+import { useRouter } from 'next/router';
+
+import { isLoggedIn } from '../../services/auth';
+
+const withAuth = (Component) => (props) => {
+  const router = useRouter();
+
+  useEffectOnce(() => {
+    if (!isLoggedIn()) {
+      router.push('/login');
+    }
+  });
+
+  return <Component {...props} />;
+};
+export default withAuth;
+```
+
+### WithModal
+
+Another HOC is `WithModal`, which pushes modalName inside props that can be used inside components.
+
+```jsx
+// /components/higherOrderComponents/withModal.tsx
+import React from 'react';
+import { useRouter } from 'next/router';
+
+import useModalName from '../../customEffects/useModalName';
+
+const withModal = (Component) => (props) => {
+  const modalName = useModalName();
+
+  return <Component modalName={modalName} {...props} />;
+};
+export default withModal;
+```
+
+### Page with Modal
+
+Example of how can we use modal on a page:
+
+```jsx
+// /pages/with-hoc-modal.tsx
+import React, { useState } from 'react';
+import Link from 'next/link';
+import Router from 'next/router';
+import modals from '../components/modals';
+
+import WithModal from '../components/higherOrderComponents/WithModal';
+import Layout from '../components/layouts/Layout';
+
+const pageUrlSlug = 'with-hoc-modal';
+
+export const withHocModalPage = (props: {
+  modalName: string,
+  onModalSubmit: (email: string) => void,
+  onModalClose: () => void,
+}) => {
+  const [name, setName] = useState(0);
+  React.useEffect(() => {
+    setName(name || Date.now());
+  }, []);
+
+  const onModalClose = () => {
+    Router.push(`/${pageUrlSlug}`);
```

Comment:
this is not good pattern because if you have pagination query parameter in url `?page=1&modal=newsletter` this line will remove it... we could use https://developer.mozilla.org/en-US/docs/Web/API/URL to deconstruct url and add or replace just parts that we need without losing unwanted data...